src/ts_emit.iced

```diff
@@ -18,8 +22,99 @@ exports.TypescriptEmitter = class TypescriptEmitter extends BaseEmitter
-  run : ({infiles, outfile, json, options}) ->
-    @emit_preface infiles, json, options
-    @_code
+  convert_primitive_type : (type) ->
+    map =
+      bool : "boolean"
+      bytes : "Buffer"
+      long : "number"
+      float : "number"
+      double : "number"
+      uint : "number"
+      int : "number"
+      uint64 : "number"
+      int64 : "number"
+    map[type] or type
+
+  make_map_type : (type) ->
+    "{[key: string]: #{@emit_field_type(type.values).type}}"
+
+  emit_field_type : (t, {optionalkey} = {}) ->
+    optional = optionalkey?
+    type = if typeof(t) is 'string' then @convert_primitive_type(t)
+    else if typeof(t) is 'object'
+      if Array.isArray(t)
+        if not t[0]?
+          optional = true
+          @emit_field_type(t[1]).type
+        else
+          "ERROR"
+      else if t.type is "array"
+        @emit_field_type(t.items).type + "[]"
+      else if t.type is "map"
+        @make_map_type t
+      else "ERROR"
+    else "ERROR"
+    { type, optional }
+
+  emit_typedef : (type) ->
+    @output "export type #{type.name} = #{@emit_field_type(type.typedef).type}"
+    @output ""
+
+  emit_imports : ({imports}) ->
+    imports = _.uniqWith imports, _.isEqual
+    imports_to_emit = imports.filter((imp) -> imp.path.indexOf('/') >= 0)
+
+    for {import_as, path} in imports_to_emit
+      @output "import * as #{import_as} from '#{path}'" if import_as
+    @output "" if imports_to_emit.length > 0
+
+
+  emit_fixed : (type) ->
+    @output "export type #{type.name} = string | null"
+
+  emit_field : ({name, type, optionalkey}) ->
+    {type, optional} = @emit_field_type(type, {optionalkey})
+    @output "#{name}#{if optional then '?' else ''}: #{type}"
+
+  emit_record : (record) ->
+    @output "export type #{record.name} = {"
+    @tab()
+    for f in record.fields
+      @emit_field
+        name : f.name
+        type : f.type
+        optionalkey : f.optional
+    @untab()
+    @output "}"
+
+  emit_enum : (type) ->
+    @output "export enum #{type.name} {"
+    @tab()
+    for s, _ in type.symbols
+      [e_name..., e_num] = s.split("_")
+      e_name = e_name.join("_")
+      @output "#{e_name} = #{e_num},"
+    @untab()
+    @output "}"
+
+  emit_variant : (type) ->
+    cases = type.cases
+      .map((type_case) =>
+        if type_case.label.def then return null
+        bodyType = switch
+          when type_case.body == null then 'null'
+          when typeof type_case.body == 'string' then @convert_primitive_type type_case.body
+          when type_case.body.type == 'array' then @convert_primitive_type(type_case.body.items) + '[]'
+          else ''
+        bodyStr = if type_case.body then ", #{type_case.label.name}: #{bodyType} | null" else ''
+        "{ #{type.switch.name}: #{type.switch.type}.#{type_case.label.name}#{bodyStr} }")
+      .filter(Boolean)
+    @output "export type #{type.name} = #{cases.join(" | ")}"
+    @output ""
```

Comment:
In `emit_variant` wouldn't `return null` chuck out the default type? `else ''` looks like the emitter doesn't recongize the input so should be "ERROR"